src/libsystemd/sd-journal/journal-file.c

```diff
@@ -1920,16 +1921,29 @@ static int journal_file_append_entry_internal(
                 return r;
 
         if (boot_id)
-                f->header->boot_id = *boot_id;
+                b = *boot_id;
+        else {
+                r = sd_id128_get_boot(&b);
+                if (r < 0)
+                        return r;
+        }
 
         if (JOURNAL_HEADER_COMPACT(f->header)) {
+                if (sd_id128_is_null(f->header->boot_id))
+                        f->header->boot_id = b;
+
+                if (!sd_id128_equal(f->header->boot_id, b))
+                        return -EXDEV;
+
                 o->entry.compact.seqnum = htole32(journal_file_new_entry_seqnum(f, seqnum) - le64toh(f->header->head_entry_seqnum));
                 o->entry.compact.realtime = htole64(ts->realtime);
                 o->entry.compact.monotonic = htole64(ts->monotonic);
-                o->entry.compact.boot_id = f->header->boot_id;
                 o->entry.compact.xor_hash = htole64(xor_hash);
                 memcpy_safe(o->entry.compact.items, items, n_items * sizeof(EntryItem));
         } else {
+                if (!sd_id128_equal(f->header->boot_id, b))
```

Comment:
this check is redundant... if it's already equal you might as well set it to b again, it won't change a thing